portable/NetworkInterface/ThirdParty/MSP432/NetworkInterface.c

```diff
@@ -0,0 +1,981 @@
+    {
+        ulTaskNotifyTakeIndexed( TX_QUEPOS_FIRST_EVENT, pdTRUE, portMAX_DELAY );
+        xQueueReceive( xQueueTX, &NIDataOutput, 0 );
+
+        /* For ICMP packets, the checksum must be zero if the network hardware computes the checksum or the buffer will not be properly sent
+         * For this driver it is required that ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM == 1 */
+        ProtocolPacket_t * pxPacket;
+        pxPacket = ( ProtocolPacket_t * ) ( NIDataOutput.pxDescriptor->pucEthernetBuffer );
+
+        if( pxPacket->xICMPPacket.xIPHeader.ucProtocol == ( uint8_t ) ipPROTOCOL_ICMP )
+        {
+            pxPacket->xICMPPacket.xICMPHeader.usChecksum = ( uint16_t ) 0u;
+        }
+
+        taskENTER_CRITICAL();
+        packetTransmit( NIDataOutput.pxDescriptor->pucEthernetBuffer, ( uint32_t ) NIDataOutput.pxDescriptor->xDataLength );
+        taskEXIT_CRITICAL();
+
+        /* wait until transmit */
+        ulTaskNotifyTakeIndexed( TX_QUEPOS_SECOND_EVENT, pdTRUE, portMAX_DELAY );
+        iptraceNETWORK_INTERFACE_TRANSMIT();
+
+        if( NIDataOutput.xReleaseAfterSend )
+        {
+            vReleaseNetworkBufferAndDescriptor( NIDataOutput.pxDescriptor );
+        }
+    }
+}
+
+
+/*  Function to output packets to the network.
+ *  Called directly by FreeRTOS. This function should not block and therefore passes
+ *  all output data to a queue.  If the function blocks the network stack can become unresponsive
+ *  so all of the output work is thereby done by another task. */
+BaseType_t xNetworkInterfaceOutput( NetworkBufferDescriptor_t * const pxDescriptor,
+                                    BaseType_t xReleaseAfterSend )
+{
+    struct NetworkInterfaceDataOut NIDataOutput;
+    BaseType_t txResp;
+
+    NIDataOutput.pxDescriptor = pxDescriptor;
+    NIDataOutput.xReleaseAfterSend = xReleaseAfterSend;
+    txResp = xQueueSend( xQueueTX, &NIDataOutput, 0 );
+
+    if( txResp == pdTRUE )
+    {
+        xTaskNotifyGiveIndexed( xTaskToNotifyEthernetTX, 0 );
+        return pdTRUE;
+    }
+
+    return pdFALSE;
+}
+
+
+/*
+ * -----------------------------------------------------------------------------------------------------
+ * Network interface UP or DOWN tasks and associated functions
+ * -----------------------------------------------------------------------------------------------------
+ */
+
+/*  Function to check and see if the network is up or down by polling a register in the Ethernet driver.
+ *  Apparently there is no other way to do this other than polling.  The register is checked for link status,
+ *  but also for remote faults and this takes into consideration a pulled Ethernet cable during RX and TX.
+ */
+bool isEMACLinkUp()
+{
+    uint8_t uci8PHYAddr = 0; /* refers to the internal PHY */
+    uint16_t check;
+
+    check = EMACPHYRead( EMAC0_BASE, uci8PHYAddr, EPHY_BMSR );
+
+    if( ( ( check & EPHY_BMSR_LINKSTAT ) == 0 ) || ( check & EPHY_BMSR_RFAULT ) )
+    {
+        return false; /* link is not up */
```

Comment:
See previous comment about rewriting with a single `return`, that would work here as well.

Reply:
Yes, the code is updated.